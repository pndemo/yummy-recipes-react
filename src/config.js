/*
* Application configurations
*/

import axios from 'axios';

// API URLs
export const baseAPIURL = 'http://yummy-recipes-apis.herokuapp.com';
export const authAPIURL = baseAPIURL.concat('/api/v1/auth/');
export const categoryAPIURL = baseAPIURL.concat('/api/v1/category/');
export const recipeAPIURL = baseAPIURL.concat('/api/v1/recipe/');

// axios instance for private API calls
const privateAxiosInstance = axios.create({
  headers: { Authorization: `Bearer ${localStorage.getItem('token')}` },
});
privateAxiosInstance.interceptors.request.use((config) => {
  if (localStorage.getItem('token') && config.headers.Authorization === 'Bearer null') {
    config.headers.Authorization = `Bearer ${localStorage.getItem('token')}`;
  }
  return config;
});

export default privateAxiosInstance;
