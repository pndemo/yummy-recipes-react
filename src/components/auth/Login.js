/*
* -- Login application component --
*  Enables a register user to login
*/

import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { authAPIURL } from '../../config';
import FooterStatic from '../common/FooterStatic';
import axios from 'axios'

// Application styling
import '../../static/App.css';

class Login extends Component {
    // App component properties
    constructor(props) {
        super(props);
        this.state = {
            isLoggedIn: localStorage.getItem('token'),
            username: '',
            password: '',
            usernameError: '',
            passwordError: '',
            error: '',
        };
    }

    // Reset error state
    resetErrorState = () => {
        this.setState({
            usernameError: '',
            passwordError: '',
            error: '',
        });
    };

    // When input changes
    onInputChanged = (event) => {
        this.setState({
            [event.target.name]: event.target.value
        })
    };

    // Handle user login
    onLoginClick = (event) => {
        event.preventDefault();
        axios.post(`${ authAPIURL }login`,
            {
                username: this.state.username,
                password: this.state.password,
        })
        .then((response) => {
            if (response.status === 200) {
                localStorage.setItem('token', response.data['access_token']);
                this.setState({
                    isLogged: true,
                })
            }
        })
        .catch((error) => {
            this.resetErrorState();
            if (error.response.status === 400) {
                if (error.response.data['username_message'] !== 'Valid') {
                    this.setState({usernameError: error.response.data['username_message']});
                }
                if (error.response.data['password_message'] !== 'Valid') {
                    this.setState({passwordError: error.response.data['password_message']});
                }
            } else if (error.response.status === 401 || error.response.status === 500) {
                this.setState({error: error.response.data['message']});
            }
        });
    };

    // Component rendering
    render() {
        if (this.state.isLogged) {
            return <Redirect to="/"/>
        }
        return (
            <div className="App-light">
                <div className="container-fluid">
                    <div className="row">
                        <div className="col-xs-12 col-sm-10 col-md-8 col-lg-6 col-sm-offset-1 col-md-offset-2 col-lg-offset-3">
                            <div className="App-header">
                                <h1 className="App-title-form">
                                    <span className="glyphicon glyphicon-fire"></span> Yummy Recipes
                                </h1>
                            </div>
                            <div className="panel panel-primary">
                                <div className="panel-heading text-center">
                                    Login
                                </div>
                                <div className="panel-body">
                                    <form onSubmit={ this.onLoginClick }>
                                        <p className="error">{ this.state.error }</p>
                                        <div className="form-group">
                                            <input
                                                className="form-control"
                                                type="text"
                                                name="username"
                                                value={ this.state.username }
                                                onChange={ this.onInputChanged }
                                                maxLength="50"
                                                placeholder="Username"/>
                                                <p className="text-left error">{ this.state.usernameError }</p>
                                        </div>
                                        <div className="form-group">
                                            <input
                                                className="form-control"
                                                type="password"
                                                name="password"
                                                value={ this.state.password }
                                                onChange={ this.onInputChanged }
                                                maxLength="50"
                                                placeholder="Password"/>
                                                <p className="text-left error">{ this.state.passwordError }</p>
                                        </div>
                                        <div className="form-group">
                                            <button type="submit" className="btn btn-primary btn-block">Login</button>
                                        </div>
                                        <p className="form-group text-center">
                                            Don't have an account? <a href="/register">Register</a>
                                        </p>
                                        <p className="form-group text-center">
                                            Forgot password? <a href="/reset_password">Reset</a>
                                        </p>
                                    </form>
                                </div>
                            </div>
                        </div>
                    </div>
                    <FooterStatic/>
                </div>
            </div>
        );
    }
}

export default Login;
