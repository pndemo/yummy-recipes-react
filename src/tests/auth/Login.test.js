import React from 'react';
import { mount, configure } from 'enzyme';
import { StaticRouter } from 'react-router';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import Login from '../../components/auth/Login';
import { authAPIURL } from '../../config';

const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');

// Set the mock adapter on default instance
const mock = new MockAdapter(axios);

// Mock login POST request to /login
mock.onPost(`${authAPIURL}login`).reply(201, { message: 'You are now logged in.' });

// Local storage
global.localStorage = {
  getItem: () => {},
  setItem: () => {},
};

// Component Login
describe('Component: Login', () => {
  configure({ adapter: new Adapter() });
  const routerComponent = mount(
  <StaticRouter location="login" context={{}}>
    <Login />
  </StaticRouter>
  );

  const loginComponent = routerComponent.find('Login');
  const usernameInput = loginComponent.find('[name="username"]');
  const passwordInput = loginComponent.find('[name="password"]');
  const login = loginComponent.instance();

  it('Display Login', () => {
    const rendered = renderer.create(
      <StaticRouter location="login" context={{}}>
        <Login />
      </StaticRouter>,
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('updates state on input change', () => {
    usernameInput.simulate('change', { target: { name: 'username', value: 'testuser' } });
    passwordInput.simulate('change', { target: { name: 'password', value: 'Bootcamp' } });
    expect(login.state.username).toEqual('testuser');
    expect(login.state.password).toEqual('Bootcamp');
  });

  it('login user', () => {
    const loginButton = loginComponent.find('[type="submit"]');
    loginButton.simulate('submit');
  });
});
