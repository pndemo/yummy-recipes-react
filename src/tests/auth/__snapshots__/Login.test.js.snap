// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Login Display Login 1`] = `
<div
  className="App-light"
>
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <div
        className="col-xs-12 col-sm-10 col-md-8 col-lg-6 col-sm-offset-1 col-md-offset-2 col-lg-offset-3"
      >
        <div
          className="App-header"
        >
          <h1
            className="App-title-form"
          >
            <span
              className="glyphicon glyphicon-fire"
            />
             Yummy Recipes
          </h1>
        </div>
        <div
          className="panel panel-primary"
        >
          <div
            className="panel-heading text-center"
          >
            Login
          </div>
          <div
            className="panel-body"
          >
            <form
              onSubmit={[Function]}
            >
              <p
                className="error"
              >
                
              </p>
              <div
                className="form-group"
              >
                <input
                  className="form-control"
                  maxLength="50"
                  name="username"
                  onChange={[Function]}
                  placeholder="Username"
                  type="text"
                  value=""
                />
                <p
                  className="text-left error"
                >
                  
                </p>
              </div>
              <div
                className="form-group"
              >
                <input
                  className="form-control"
                  maxLength="50"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <p
                  className="text-left error"
                >
                  
                </p>
              </div>
              <div
                className="form-group"
              >
                <button
                  className="btn btn-primary btn-block"
                  type="submit"
                >
                  Login
                </button>
              </div>
              <p
                className="form-group text-center"
              >
                Don't have an account? 
                <a
                  href="/register"
                >
                  Register
                </a>
              </p>
              <p
                className="form-group text-center"
              >
                Forgot password? 
                <a
                  href="/reset_password"
                >
                  Reset
                </a>
              </p>
            </form>
          </div>
        </div>
      </div>
    </div>
    <footer
      className="App-footer-static"
    >
      <p
        className="App-copyright"
      >
        Â© 2018 Yummy Recipes LLC
      </p>
    </footer>
  </div>
</div>
`;
